sp/src/game/client/mapbase/vscript_vgui.cpp

```diff
@@ -0,0 +1,1871 @@
+
+		Check( OnTick );
+		Check( OnScreenSizeChanged );
+		Check( PerformLayout );
+
+		g_pScriptVM->RaiseException("invalid callback");
+
+#undef Check
+	}
+};
+
+//--------------------------------------------------------------
+//--------------------------------------------------------------
+
+class CScript_Label : public Label, public CScriptVGUIPanel
+{
+public:
+	CScript_Label(Panel *parent, const char *name, const char *text)
+		: Label( parent, name, text )
+	{}
+};
+
+//--------------------------------------------------------------
+//--------------------------------------------------------------
+
+
+//==============================================================
+//==============================================================
+
+//--------------------------------------------------------
+// Script objects
+//--------------------------------------------------------
+class CScript_Panel_Helper : public IScript_Panel< CScript_Panel >
+{
+	DECLARE_SCRIPTVGUI_HELPER( CScript_Panel )
+
+public:
+	void SetCallback( const char* cb, HSCRIPT fn )
+	{
+		_base->SetCallback( cb, fn );
+	}
+};
+
+BEGIN_SCRIPTDESC_VGUI( Panel )
+	DEFINE_SCRIPTFUNC( SetCallback, "" )
+END_SCRIPTDESC()
+
+//--------------------------------------------------------------
+//--------------------------------------------------------------
+
+class CScript_Label_Helper : public IScript_Label< CScript_Label >
+{
+	DECLARE_SCRIPTVGUI_HELPER_DEFAULT_TEXT( CScript_Label )
+
+public:
+};
+
+BEGIN_SCRIPTDESC_VGUI( Label )
+END_SCRIPTDESC()
+
+//--------------------------------------------------------------
+//--------------------------------------------------------------
+
+
+//==============================================================
+//==============================================================
+
+
+class CScriptVGUI : public CAutoGameSystem
+{
+public:
+	void LevelShutdownPostEntity();
+	void Shutdown();
+
+public:
+	HSCRIPT Create( const char* type, HSCRIPT parent, const char* name, int root );
+
+} script_vgui;
+
+BEGIN_SCRIPTDESC_ROOT_NAMED( CScriptVGUI, "IVGui", SCRIPT_SINGLETON )
+	DEFINE_SCRIPTFUNC( Create, SCRIPT_HIDE )
+END_SCRIPTDESC()
+
+
+HSCRIPT CScriptVGUI::Create( const char* type, HSCRIPT parent, const char* name, int root )
+{
+	// Init root panel on first script panel creation.
+	if ( !g_pScriptRootPanel )
+	{
+		g_pScriptRootPanel = new CScriptRootPanel();
+	}
+
+#define Check( _name )\
+	if ( FStrEq( type, #_name ) )\
+	{\
+		CScript_##_name##_Helper *helper = new CScript_##_name##_Helper;\
+		HSCRIPT ret = g_pScriptVM->RegisterInstance( helper );\
+		helper->CreateFromScript( (HSCRIPT)parent, name, ret, root );\
+		g_ScriptPanels.AddToTail( helper );\
+		return ret;\
+	}
+
+	Check( Panel );
+	Check( Label );
+	//Check( Button );
+
+	g_pScriptVM->RaiseException("invalid vgui panel type");
+	return NULL;
+
+#undef Check
+}
+
+void CScriptVGUI::LevelShutdownPostEntity()
+{
+	FOR_EACH_VEC( g_ScriptPanels, i )
+	{
+		g_ScriptPanels[i]->Destroy();
```

Comment:
Maybe `IScriptVGUIObject::~IScriptVGUIObject()` should call `Destroy()`.

Reply:
It causes a build error. I didn't think it was important enough to look into.

Reply:
Ah, calling virtual functions in the destructor (also constructor) works like static dispatch, so the derived methods wouldn't be invoked.